source/logrepl/internal/subscription.go

```diff
@@ -27,125 +27,119 @@ import (
 	"github.com/jackc/pgx/v5/pgproto3"
 )
 
-const (
-	pgDuplicateObjectErrorCode = "42710"
-	pgOutputPlugin             = "pgoutput"
-)
+const pgOutputPlugin = "pgoutput"
 
 // Subscription manages a subscription to a logical replication slot.
 type Subscription struct {
-	ConnConfig    pgconn.Config
 	SlotName      string
 	Publication   string
 	Tables        []string
 	StartLSN      pglogrepl.LSN
 	Handler       Handler
 	StatusTimeout time.Duration
+	TXSnapshotID  string
+
+	conn *pgconn.PgConn
+
+	stop context.CancelFunc
 
-	stop    context.CancelFunc
 	ready   chan struct{}
 	done    chan struct{}
 	doneErr error
 
-	// cleanup is the function that gets called on teardown.
-	// Cleanup functions that get added here on initialization act as deferred
-	// functions.
-	cleanup func(ctx context.Context) error
-
 	walWritten pglogrepl.LSN
 	walFlushed pglogrepl.LSN
 }
 
 type Handler func(context.Context, pglogrepl.Message, pglogrepl.LSN) error
 
 func NewSubscription(
-	config pgconn.Config,
+	ctx context.Context,
+	conn *pgconn.PgConn,
 	slotName,
 	publication string,
 	tables []string,
 	startLSN pglogrepl.LSN,
 	h Handler,
-) *Subscription {
+) (*Subscription, error) {
+	result, err := pglogrepl.CreateReplicationSlot(
```

Comment:
One thing that caught my eye is that `NewSubscription`, a constructor, has a side effect other than creating the `Subscription` object. If you ask me, it's a code smell.